ruma-client-api/src/r0/session/get_login_types.rs

```diff
@@ -40,51 +48,345 @@ impl Response {
+    pub identity_providers: Option<Vec<IdentityProvider>>,
+}
+
+impl SsoLoginType {
+    /// Creates a new `PasswordLoginType`.
+    pub fn new() -> Self {
+        Self {
+            #[cfg(feature = "unstable-pre-spec")]
+            identity_providers: None,
+        }
+    }
+}
+
+impl Default for SsoLoginType {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// An SSO login identity provider.
+#[cfg(feature = "unstable-pre-spec")]
+#[cfg_attr(docsrs, doc(cfg(feature = "unstable-pre-spec")))]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct IdentityProvider {
+    /// The ID of the provider.
+    id: String,
+
+    /// The display name of the provider.
+    name: String,
+
+    /// The icon for the provider.
+    icon: Option<MxcUri>,
+
+    /// The brand identifier for the provider.
+    brand: Option<IdentityProviderBrand>,
+}
+
+#[cfg(feature = "unstable-pre-spec")]
+impl IdentityProvider {
+    /// Creates an `IdentityProvider` with the given `id` and `name`.
+    pub fn new(id: &str, name: &str) -> Self {
+        Self { id: id.into(), name: name.into(), icon: None, brand: None }
+    }
+}
+
+/// An SSO login identity provider brand identifier.
+///
+/// This uses the unstable prefix in
+/// [MSC2858](https://github.com/matrix-org/matrix-doc/pull/2858).
+#[cfg(feature = "unstable-pre-spec")]
+#[cfg_attr(docsrs, doc(cfg(feature = "unstable-pre-spec")))]
+#[derive(Clone, Debug, PartialEq, Eq, StringEnum)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum IdentityProviderBrand {
+    /// The [Apple] brand.
+    ///
+    /// [Apple]: https://developer.apple.com/design/human-interface-guidelines/sign-in-with-apple/overview/buttons/
+    #[ruma_enum(rename = "org.matrix.apple")]
+    Apple,
+
+    /// The [Facebook](https://developers.facebook.com/docs/facebook-login/web/login-button/) brand.
+    #[ruma_enum(rename = "org.matrix.facebook")]
+    Facebook,
+
+    /// The [GitHub](https://github.com/logos) brand.
+    #[ruma_enum(rename = "org.matrix.github")]
+    Github,
+
+    /// The [GitLab](https://about.gitlab.com/press/press-kit/) brand.
+    #[ruma_enum(rename = "org.matrix.gitlab")]
+    Gitlab,
+
+    /// The [Google](https://developers.google.com/identity/branding-guidelines) brand.
+    #[ruma_enum(rename = "org.matrix.google")]
+    Google,
+
+    /// The [Twitter] brand.
+    ///
+    /// [Twitter]: https://developer.twitter.com/en/docs/authentication/guides/log-in-with-twitter#tab1
+    #[ruma_enum(rename = "org.matrix.twitter")]
+    Twitter,
+
+    /// A custom brand.
     #[doc(hidden)]
     _Custom(String),
 }
 
+/// A custom login payload.
+#[doc(hidden)]
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct CustomLoginType {
+    /// A custom type
+    ///
+    /// This field is named `type_` instead of `typee` because the latter is a reserved
```

Comment:
No, `typee` is not a reserved keyword 😄 

```suggestion
    /// This field is named `type_` instead of `type` because the latter is a reserved
```